### PR TITLE
Improve API documentation + deploy documentation site

### PR DESCRIPTION
**Pull request by vgoh1**
# Description 
The function documentation for publications and users controllers have been refined - more detailed parameter and return code descriptions. Added documentation for middleware. Using jsdoc to generate the pages, the documentation has been deployed to GitHub pages at https://researchify.github.io/. A workflow action has been added so that whenever a push to main happens, it triggers the deployment of the site (including generating the docs again from main).

# Type of change 
- [ ] Bug fix
- [x] New feature
- [ ] Refactoring
- [x] Documentation 

# Problem
A documentation site may help the rest of the team understand how the APIs are used. The documentation needed to improve first.

# Solution description
Improved function documentation is self explanatory.
Added a readme in the api folder that is used as the homepage of the documentation site. 
Added a stage to the workflow yml (trigger-deploy-doc) that remotely triggers the workflow to generate the docs and deploy it: https://github.com/Researchify/researchify.github.io/blob/main/.github/workflows/main.yml (note: hardcoded to generate docs from this branch right now, will change after this has been merged into main)

The main.yml in the docs site repo consists of checking out the main Researchify repo, and then the docs repo. After the docs are generated from the main Researchify repo, they get moved to the docs repo, committed and then pushed. There is no deploy stage because GH pages automatically sets up the pages from the repo. The deployment involves overwriting all of the files in the repo with the generated ones so it's safer to have it in a separate repo.

# Tests 
See https://researchify.github.io/ for the live site
Successful trigger after pushing to this branch (vgoh1/api-documentation):
image

If you want to test it out yourself, do a dummy push in my branch to see it trigger in actions for this repo and then the docs repo
image

# Relevant document/ link (if any) 
N/A

# Risk
Low risk because the docs are hosted in another repo so it won't affect the main repo code.

### CODE REVIEW

**Comment by rsun0013** [comment]
Perhaps the methods to access the Api endpoints could be listed as well

**Review by rsun0013 — APPROVED** [review]

**Comment by vgoh1** [comment]
> Perhaps the methods to access the Api endpoints could be listed as well

@rsun0013 What do you mean? As in what to use to hit the endpoints or the http methods?

**Review by pmyip3 — APPROVED** [review]

File: api/jsdoc_config.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+    "plugins": [],
+    "recurseDepth": 10,
+    "source": {
+        "exclude": ["./api/src/controllers/foo.js"]
```
Comment (aksh0001) [inline api/jsdoc_config.json:5]: nitpick: this file will be removed in this PR https://github.com/Researchify/Researchify/pull/65

File: api/package.json [hunk 2]
```diff
@@ -30,5 +31,8 @@
   },
   "jest": {
     "testEnvironment": "node"
+  },
+  "script": {
+    "generate-docs": "node_modules/.bin/jsdoc -c jsdoc_config.json"
```
Comment (aksh0001) [inline api/package.json:36]: curious: is there a difference between `script` and the `scripts` prop being used above? i.e. https://docs.npmjs.com/cli/v7/using-npm/scripts

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -35,6 +35,15 @@ jobs:
...
         working-directory: ./client
+  trigger-deploy-doc:
+    name: Generate and deploy docs for researchify.github.io
+    runs-on: ubuntu-latest
+    steps:
+      - uses: peter-evans/repository-dispatch@v1
+        with:
+          token: ${{ secrets.DOCS_WORKFLOW_PAT }}
+          repository: Researchify/researchify.github.io
+          event-type: trigger-deploy
```
Comment (aksh0001) [inline .github/workflows/main.yml:46]: curious: are we going to be deploying docs on commits to non-`main` branches as well?
Also, I'm wondering how this CI job relates to the `script` defined for the docs? https://github.com/Researchify/Researchify/blob/be7004aaaa2de6f6040542d2c183ced1a4fb95dc/api/package.json#L35
More specifically, where are we generating and deploying the docs here?

Comment (aksh0001) [reply .github/workflows/main.yml:46]: I see that the github pages docs repo also has a CI script: https://github.com/Researchify/researchify.github.io/blob/main/.github/workflows/main.yml
I understand now. Is it worth documenting the relationship between these two CI scripts somewhere to make it easier to understand?

**Review by aksh0001 — APPROVED** [review]
Great job! Thanks for updating the Readme with the doc generation instructions!

File: api/jsdoc_config.json (hunk 1 above)
Comment (vgoh1) [reply api/jsdoc_config.json:5]: Ok, will remove this line then

File: .github/workflows/main.yml (hunk 3 above)
Comment (vgoh1) [reply .github/workflows/main.yml:46]: 1. Good catch - I plan to add a condition to only trigger if its main branch before this is merged. I don't think the script / scripts are related.
2. Yup I think I'll document the relationship in the readme of the docs repo

**Pull request merged**